homeassistant/components/alexa/auth.py

```diff
@@ -0,0 +1,148 @@
+        # be an access token request
+
+        lwa_params = {
+            "grant_type": "authorization_code",
+            "code": accept_grant_code,
+            "client_id": self.client_id,
+            "client_secret": self.client_secret
+        }
+        _LOGGER.debug("Calling LWA to get the access token (first time), "
+                      "with: %s", json.dumps(lwa_params))
+
+        return await self._async_request_new_token(lwa_params)
+
+    async def async_get_access_token(self):
+        """Perform access token or token refresh request."""
+        if self._prefs is None:
+            await self.async_load_preferences()
+
+        if self.is_token_valid():
+            _LOGGER.debug("Token still valid, using it.")
+            return self._prefs[STORAGE_ACCESS_TOKEN]
+
+        if self._prefs[STORAGE_REFRESH_TOKEN] is None:
+            _LOGGER.debug("Token invalid and no refresh token available.")
+            return None
+
+        lwa_params = {
+            "grant_type": "refresh_token",
+            "refresh_token": self._prefs[STORAGE_REFRESH_TOKEN],
+            "client_id": self.client_id,
+            "client_secret": self.client_secret
+        }
+
+        _LOGGER.debug("Calling LWA to refresh the access token.")
+        return await self._async_request_new_token(lwa_params)
+
+    @callback
+    def is_token_valid(self):
+        """Check if a token is already loaded and if it is still valid."""
+        if not self._prefs[STORAGE_ACCESS_TOKEN]:
+            return False
+
+        expire_time = datetime.strptime(self._prefs[STORAGE_EXPIRE_TIME],
+                                        DATE_FORMAT)
+        preemptive_expire_time = expire_time - timedelta(
+            seconds=PREEMPTIVE_REFRESH_TTL_IN_SECONDS)
+
+        return datetime.utcnow() < preemptive_expire_time
+
+    async def _async_request_new_token(self, lwa_params):
+
+        try:
+            session = aiohttp_client.async_get_clientsession(self.hass)
+            with async_timeout.timeout(DEFAULT_TIMEOUT, loop=self.hass.loop):
+                response = await session.post(LWA_TOKEN_URI,
+                                              headers=LWA_HEADERS,
+                                              data=lwa_params,
+                                              allow_redirects=True)
+
+        except (asyncio.TimeoutError, aiohttp.ClientError):
+            _LOGGER.error("Timeout calling LWA to get auth token.")
+            return None
+
+        response_text = await response.text()
+
+        _LOGGER.debug("LWA response header: %s", response.headers)
+        _LOGGER.debug("LWA response status: %s", response.status)
+        _LOGGER.debug("LWA response body  : %s", response_text)
+
+        if response.status != 200:
+            _LOGGER.error("Error calling LWA to get auth token.")
+            return None
+
+        access_token = json.loads(response_text)["access_token"]
```

Comment:
This is very inefficient, loading the json on every line. Just load it as JSON as I suggested above.